pip/req/req_install.py

```diff
@@ -421,6 +421,14 @@ def run_egg_info(self):
                     self.pkg_info()["Version"],
                 ]))
             self._correct_build_location()
+        else:
+            metadata_name = canonicalize_name(self.pkg_info()["Name"])
+            if canonicalize_name(self.req.project_name) != metadata_name:
+                raise InstallationError(
+                    'Running setup.py (path:%s) egg_info for package %s '
```

Comment:
I know this after merge, but thinking the message should mention the url (if present) with the #egg fragment, or otherwise I'm not sure people will connect the dots that the problem is simply due to the fragment being wrong.